File: src/awaitable_event/awaitable_event_workflow.erl
-module(awaitable_event_workflow).

% elp:ignore W0012 W0040
-moduledoc false.

-export([
    execute/2
]).

-include_lib("temporal_sdk/include/workflow.hrl").

execute(_Context, _Input) ->
    {activity_start, AId} =
        AStart = start_activity(echo_activity, [[~"Activity completed."]], [
            {task_queue, "awaitable_event"},
            {awaitable_event, start}
        ]),
    {AwaitStatus, #{}} = await(AStart, 1_000),
    #{is_replaying := IsReplaying} = workflow_info(),
    case {IsReplaying, AwaitStatus} of
        {true, _} ->
            ok;
        {false, noevent} ->
            io:fwrite(
                "WARN: Activity not started within 1000 msec timeout or nonexistent activity.~n", []
            );
        {false, ok} ->
            io:fwrite("INFO: Activity started within 1000 msec timeout.~n", [])
    end,
    AClose = {activity, AId},
    #{state := completed, result := [[AR]]} = wait(AClose),
    io:fwrite("~p~n", [AR]).

File: lib/awaitable_event/awaitable_event_workflow.ex
defmodule AwaitableEvent.Workflow do
  use TemporalSdk.Workflow
  @moduledoc false

  @impl true
  def execute(_context, _input) do
    {:activity_start, a_id} =
      a_start =
      start_activity(EchoActivity, [["Activity completed."]],
        task_queue: "awaitable_event",
        awaitable_event: :start
      )

    {await_status, %{}} = await(a_start, 1_000)
    %{is_replaying: is_replaying} = workflow_info()

    case {is_replaying, await_status} do
      {true, _} ->
        :ok

      {false, :noevent} ->
        IO.puts("WARN: Activity not started within 1000 msec timeout or nonexistent activity.")

      {false, :ok} ->
        IO.puts("INFO: Activity started within 1000 msec timeout.")
    end

    a_close = {:activity, a_id}
    %{state: :completed, result: [[r]]} = wait(a_close)
    IO.puts(r)
  end
end
